File: chamber-library/src/main/java/com/dtp/DataConnection.java
package com.dtp;

import java.util.List;

/**
 * Created by ner on 4/6/17.
 */

public class DataConnection {
    //TODO Implement upsert
    //TODO Implement Async getAndClose and doAndClose
    //TODO Implement Async with RxJava
    //TODO Implement table watchers with RxJava

    private static DatabaseManager databaseManager;

    private final Database database;

    private DataConnection(Database database) {
        this.database = database;
    }

    public static void init(DatabaseManager databaseManager) {
        DataConnection.databaseManager = databaseManager;
    }

    public static void doAndClose(DoAndClose block) {
        block.doAndClose(openConnection());
    }

    public static <T> T getAndClose(GetAndCloseBlock<T> block) {
        return block.getAndClose(openConnection());
    }

    private static DataConnection openConnection() {
        return new DataConnection(databaseManager.beginTransaction());
    }

    public void insert(DataTable dataTable) {
        Database database = databaseManager.beginTransaction();

        insert(dataTable, database);

        databaseManager.endTransaction();
    }

    public void insertAll(List<DataTable> dataTables) {
        Database database = databaseManager.beginTransaction();

        for (DataTable dataTable : dataTables) {
            insert(dataTable, database);
        }

        databaseManager.endTransaction();
    }

    private void insert(DataTable dataTable, Database database) {
        database.insert(dataTable);
    }

    public void update(DataTable dataTable) {
        Database database = databaseManager.beginTransaction();

        update(dataTable, database);

        databaseManager.endTransaction();
    }

    public void updateAll(List<DataTable> dataTables) {
        Database database = databaseManager.beginTransaction();

        for (DataTable dataTable : dataTables) {
            update(dataTable, database);
        }

        databaseManager.endTransaction();
    }

    private void update(DataTable dataTable, Database database) {
        database.update(dataTable);
    }

    public void delete(DataTable dataTable) {
        Database database = databaseManager.beginTransaction();

        delete(dataTable, database);

        databaseManager.endTransaction();
    }

    public void delete(List<DataTable> dataTables) {
        Database database = databaseManager.beginTransaction();

        for (DataTable dataTable : dataTables) {
            delete(dataTable, database);
        }

        databaseManager.endTransaction();
    }

    private void delete(DataTable dataTable, Database database) {
        database.delete(dataTable);
    }

    public interface DoAndClose {
        void doAndClose(DataConnection dataConnection);
    }

    public interface GetAndCloseBlock<T> {
        T getAndClose(DataConnection dataConnection);
    }
}

File: chamber-library/src/main/java/com/dtp/DataConnection.kt
package com.dtp

import com.dtp.data_table.DataTable
import com.dtp.data_table.ParentDataTable
import com.dtp.query.Query

/**
 * Created by ner on 4/6/17.
 */

class DataConnection private constructor() {

    companion object {
        //TODO Implement upsert
        //TODO Implement Async getAndClose and doAndClose
        //TODO Implement Async with RxJava
        //TODO Implement table watchers with RxJava

        private lateinit var databaseManager: DatabaseManager

        fun init(databaseManager: DatabaseManager) {
            DataConnection.databaseManager = databaseManager
        }

        fun createDataStore(): DataStoreIn {
            return databaseManager.createInDataStore()
        }

        fun doAndClose(block: (DataConnection) -> Unit) {
            block(openConnection())
        }

        fun <T> getAndClose(block: (DataConnection) -> T): T {
            return block(openConnection())
        }

        private fun openConnection(): DataConnection {
            return DataConnection()
        }
    }

    fun insert(dataTable: DataTable) {
        val database = databaseManager.beginTransaction()

        insert(dataTable, database)

        databaseManager.endTransaction()
    }

    fun insertAll(dataTables: List<DataTable>) {
        val database = databaseManager.beginTransaction()

        dataTables.forEach { insert(it, database) }

        databaseManager.endTransaction()
    }

    private fun insert(dataTable: DataTable, database: Database) {
        val chamberId = database.insert(dataTable)

        dataTable.chamberId = chamberId

        if (dataTable is ParentDataTable) {
            dataTable.chamberChildren.forEach {
                it.parentChamberId = dataTable.chamberId

                insert(it, database)
            }
        }
    }

    fun update(dataTable: DataTable) {
        val database = databaseManager.beginTransaction()

        update(dataTable, database)

        databaseManager.endTransaction()
    }

    fun updateAll(dataTables: List<DataTable>) {
        val database = databaseManager.beginTransaction()

        dataTables.forEach { update(it, database) }

        databaseManager.endTransaction()
    }

    private fun update(dataTable: DataTable, database: Database) {
        database.update(dataTable)

        if (dataTable is ParentDataTable)
            dataTable.chamberChildren.forEach { update(it, database) }
    }

    fun delete(dataTable: DataTable) {
        val database = databaseManager.beginTransaction()

        delete(dataTable, database)

        databaseManager.endTransaction()
    }

    fun delete(dataTables: List<DataTable>) {
        val database = databaseManager.beginTransaction()

        dataTables.forEach { delete(it, database) }

        databaseManager.endTransaction()
    }

    private fun delete(dataTable: DataTable, database: Database) {
        database.delete(dataTable)

        if (dataTable is ParentDataTable)
            dataTable.chamberChildren.forEach { delete(it, database) }
    }

    fun <T : DataTable> findFirst(itemBuilder: ItemBuilder<T>, query: Query): T? {
        val database = databaseManager.beginTransaction()

        val item = database.findFirst(itemBuilder, query)

        databaseManager.endTransaction()

        return item
    }

    fun <T : DataTable> findAll(itemBuilder: ItemBuilder<T>, query: Query): List<T> {
        val database = databaseManager.beginTransaction()

        val items = database.findAll(itemBuilder, query)

        databaseManager.endTransaction()

        return items
    }
}
